### ISSUE TITLE
Type guards in the setter methods in the PDB module

### ISSUE THREAD

**Issue by kjczarne** [biopython/biopython#4836]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```
3.10.0 (default, Nov 10 2021, 13:28:23) [Clang 12.0.0 ]
CPython
macOS-14.0-arm64-arm-64bit
1.81
```

### Expected behaviour

With the following snippet of code:

```python
# For each of the CA atoms in the target structure, we can now update the coordinates:
for idx, atom in enumerate(target_atoms):
    if 'CA' in atom.get_id():
        atom.set_coord(superimposed[idx])
```

I would expect a type error. I just spent a good 30 min scratching my head why my downstream code fails and it turns out, in the above snippet I'm passing the entire `Atom` object as coords. The correct code would be:

```python
# For each of the CA atoms in the target structure, we can now update the coordinates:
for idx, atom in enumerate(target_atoms):
    if 'CA' in atom.get_id():
        atom.set_coord(superimposed[idx].coord)
```

### Actual behaviour

No error is thrown.

### Steps to reproduce

Using any arbitrary object in the `.set_coord()` method is permitted. There is not type checking performed. Only downstream, e.g. when the coordinates are getting unpacked as `x, y, z = atom.coords` does one realize that the wrong object has been set as a coordinate tuple.

**I am willing to work on this!** It is a UX aspect, not necessarily critical to the core capabilities of Biopython but it would save folks a lot of time to have some type bounds. I just need info whether the maintainers of Biopython would be interested in having this part improved and whether anybody is already working on this before I invest my time in this.

Thanks!

**Comment by michaelfm1211** [comment]
Glancing over [Bio/PDB/Atom.py](https://github.com/biopython/biopython/blob/master/Bio/PDB/Atom.py), it looks like none of the setters (lines 307-361) perform type checking. We should probably add type checks on all of these. The types are all annotated in the constructor so I imagine this should be an easy change.

@peterjc @mdehoon Thoughts on this? Is there any reason these setters shouldn't have type checks?

**Comment by flaar94** [comment]
I think adding type annotations to them would probably be valuable. There's some computation cost to runtime type guards, but as long as you aren't looping over millions of them it's probably fine. Also, you can just set the attributes directly if you want to do it quicker and avoid the type guard. I guess that code is ancient: from 2002, which might explain why they implemented so many getters/setters

**Comment by peterjc** [comment]
Yes, this is quite old code - it may be straightforward to modernize the property definitions too.

PDB files can certainly have tens of thousands of atoms, but I think millions exceeds the column based limits (from memory - I've not checked the format definition recently). However, the newer structural formats ought to cope with even more atoms. I'm not sure what the current record holder is.

**Comment by flaar94** [comment]
OK, then I don’t think we need to worry about computation cost. For reference, an isinstance call takes in the ballpark of 1 second for 10 million calls, so anything on the order of tens of thousands should be lost in the background noise

**Comment by kjczarne** [comment]
Sounds good. I suppose I can add the type guards in a PR and document the `setattr` trick in case anyone wants to avoid runtime type-checking overhead.

**Comment by michaelfm1211** [comment]
@kjczarne If you'd like, I can add these type checks to #4841 to avoid any trouble with possible merge conflicts. If you'd rather open a separate PR that's fine too though.
